### DISCUSSION TITLE
BaseFeature doesn't implement IFeature

### DISCUSSION THREAD

**Discussion by charlenni** [Mapsui/Mapsui#2399] [Development]
What is the reason, that BaseFeature doesn't implement the IFeature interface?

I'm working at the moment on the bug in #2389. This could be solved by changing ConcurrentBag to something else (must be thread save) or by adding a ZOrder or ZIndex to Feature. I would have added it to IFeature, but found, that BaseFeature doesn't implement IFeature and thought about why this is the case.

For me, IFeature is the interface and BaseFeature implement this interface by filling the Id handling and so on, which is needed for all features. The only method, that BaseFeature doesn't implement of IFeature is `void ClearRenderedGeometry() { } // default implementation` which could implemented as `abstarct` or `virtual`.

**Comment by pauldendulk** [comment]
oh I noticed last week that ClearRenderedGeometry() was still there. It should be removed entirely, because we do not cache in the feature anymore, only in the IRenderCache. So, that is already an improvement. 

About inheriting from IFeature. I am not sure if this is better in general, sometimes you want a number of smaller base classes that can be composed to support one interface. In this case it confused you, and you will probably not be the only one, so adding the IFeature will help in that respect. So, let's do it.

Did you have any practical problem caused by BaseFeature not being derived from IFeature?

This is also an opportunity to say that I am not a fan of the old fashioned OO approach. I have seen too many confusing big inheritance trees, and big classes. I prefer small classes, composition, and a functional approached (immutable records and separate pure functions to operate on them).

**Comment by charlenni** [comment]
Foremost: no, I have no practical problem with it. I think about creating a GroupingLayer, which could hold child layers.

I found only, that BaseLayer has the same properties as ILayer, but doesn't implement ILayer and want to know the reason.
